### DISCUSSION TITLE
ORNO OR-WE-504 Modbus Energy Meter support.

### DISCUSSION THREAD

**Discussion by laurentppol** [arendst/Tasmota#15541] [Ideas]
Hi there,  
is it already possible to use ORNO OR-WE-504 (single phase, Modbus on RS485) energy meter with Tasmota? I have seen there is support for ORNO OR-WE-517 meter.
Polish distributor provided me register map (in *..xlsx  format).
I was able to read values using qmodbus.

address	Value	Holding Registers from  0 to 15. 
0	Voltage	0 Napięcie (0,1V), 
1	current	1 Natężenie (0,1A), 
2	frequency	2 Częstotliwość (0,1Hz), 
3	active power	3 Moc czynna (1W), 
4	reactive power	4 Moc bierna (1var), 
5	apparent power	5 Moc pozorna (1VA), 
6	power factor	6 Współczynnik mocy (1000), 
7,8	active energy	7 i 8 Energia czynna (1Wh), Decimal Long - little endian 
9,A	reactive energy	9 i A Energia bierna (1varh), Decimal Long - little endian 
E	baud rate	E Prędkość transmisji: 
		1-1200 
		2-2400 
		3-4800 
		4-9600 
F	address	F Adres

Is this enough?

[OR-WE-504 RS485-modbus communication protocol.pdf](https://github.com/arendst/Tasmota/files/8629339/OR-WE-504.RS485-modbus.communication.protocol.pdf)
[register OR-WE-504.xlsx](https://github.com/arendst/Tasmota/files/8629340/register.OR-WE-504.xlsx)

Looking at WE-517 driver code seems mine would be a subset as 517 is 3 phase.

**Comment by laurentppol** [comment]
Have askesd ORNO representative to send same files about 517

W.P.

**Comment by gemu2015** [comment]
check the examples in the docs https://tasmota.github.io/docs/Smart-Meter-Interface/
and find the most similar modbus meter.
then modify the descriptor to match your meter.

**Reply by laurentppol** [reply]
Brrrr, looks as cryptic as FORTH.
I think I'll try to understand WE517 code, get transmission details for this and compare.

Can onyone tell me, what is "meter driver" (ex [xnrg_17_ornowe517.ino](https://github.com/arendst/Tasmota/blob/development/tasmota/xnrg_17_ornowe517.ino)) responsibility?
Only decode payload from serial or something else?

How is serial data handled?
And WWW displaying / MQTT?
[RTFM, but point me to right ;)]

**Comment by laurentppol** [comment]
Here: https://github.com/gituser-rk/orno-modbus-mqtt/blob/master/Register%20description%20OR-WE-514%26OR-WE-515.pdf is someting.
But in Tasmota driver there is float invoked. Why? As what I can read from datasheets all values are word or int32?

**Reply by laurentppol** [reply]
https://github.com/volkszaehler/mbmd/issues/127#issuecomment-637161974
Looks like 504 has 8 bit register addresses, and 517 has 16 bit.
I have an 504 and I was able to get reasonable readings using QModbus. (0x01 address, "read  holding register" command).

**Comment by laurentppol** [comment]
Ooops, my fault. But found ModbusRTU doc in Polish. Will be easier.

**Comment by laurentppol** [comment]
Ok, have tried to add WE504 to WE517 driver.
https://github.com/laurentppol/Tasmota/blob/development/tasmota/xnrg_17_ornowe517.ino
But does not work:
- if I connect to NodeMCU with a Minicom (on USB), no WE504 module connected yet, I can only see following sequence:
0x01, (correct), 0x03, (correct), 0x01 (not correct, should be 2 bytes), 0x84, 0x0A (CRC?), and again...
Correct sequence (to which     meter respoonds in QModbus is: 01  03  00 00  00 01.
?
What have I made bad?
